### PR TITLE
Feature/event tracking changes

### PR DESCRIPTION
**Pull request by yusuftor**
## Changes in this pull request
This increments the build number to 2.4.0 and refactors the way tracking is internally performed. We now create a `SuperwallEvent` or a `UserInitiatedEvent` and use that to track a number of events. That gets sent off to the `track(_:)` function to be sent back to the server/delegate.

Also fixes a number off linear issues:
- [SW-865](https://linear.app/superwall/issue/SW-865/implement-missing-automatically-tracked-events): Adds missing internally tracked events
- [SW-692](https://linear.app/superwall/issue/SW-692/create-app-install-event-in-sdk): Adds the `app_install` event. Note: this will get fired again if someone deletes and redownloads the app.
- [SW-854](https://linear.app/superwall/issue/SW-854/create-a-dedicated-deep-link-function): Adds a dedicated deep link function `handleDeepLink(_:)`.
- [SW-860](https://linear.app/superwall/issue/SW-860/add-triggername-to-trigger-fire-event): Adds the trigger name to a `trigger_fire` event (thanks Justin for help on that).

### Checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I added tests, an experiment, or detailed why my change isn't tested.
- [x] I added an entry to the `CHANGELOG.md` for any breaking changes, enhancements, or bug fixes.
- [x] I have reviewed the [contributing guide](https://github.com/superwall-me/paywall-ios/tree/master/.github/CONTRIBUTING.md)

### CODE REVIEW

File: Examples/SuperwallUIKitExample/SuperwallUIKitExample/Services/PaywallService.swift [hunk 1]
```diff
@@ -23,8 +23,8 @@ final class PaywallService {
     )
   }
 
-  static func trackDeepLink(url: URL) {
-    Paywall.track(.deepLinkOpen(deepLinkUrl: url))
+  static func handleDeepLink(_ url: URL) {
```
Comment (jakemor) [inline Examples/SuperwallUIKitExample/SuperwallUIKitExample/Services/PaywallService.swift:26]: Is this meant to be a breaking change? Original method was called `trackDeepLink(url: URL)`

File: Sources/Paywall/Debug/SWDebugManager.swift [hunk 2]
```diff
@@ -18,46 +18,57 @@ final class SWDebugManager {
...
+      fromUrl: deepLinkUrl,
+      withName: .paywallId
+    )
+
+    SWDebugManager.shared.launchDebugger(withPaywallId: paywallId)
   }
 
 	/// Launches the debugger for you to preview paywalls.
   ///
   /// If you call `Paywall.track(.deepLinkOpen(deepLinkUrl: url))` from `application(_:, open:, options:)` in your `AppDelegate`, this function is called automatically after scanning your debug QR code in Superwall's web dashboard.
```
Comment (jakemor) [inline Sources/Paywall/Debug/SWDebugManager.swift:50]: Update comment with new function name

File: Sources/Paywall/Paywall/Paywall View Controller/SWPaywallViewController.swift [hunk 3]
```diff
@@ -49,8 +49,7 @@ final class SWPaywallViewController: UIViewController {
 
   // Views
 	lazy var shimmerView = ShimmeringView(frame: self.view.bounds)
-  private lazy var eventHandler = WebEventHandler(delegate: self)
-  lazy var webView = SWWebView(delegate: eventHandler)
+  lazy var webView = SWWebView(delegate: self)
```
Comment (jakemor) [inline Sources/Paywall/Paywall/Paywall View Controller/SWPaywallViewController.swift:52]: I believe I needed to keep the eventHandler as a separate object due to some bug in iOS. Something relating to retain cycles / webview being deallocated (not sure, will follow up)

What was the reason for this refactor? Just general clean up?

File: Examples/SuperwallUIKitExample/SuperwallUIKitExample/Services/PaywallService.swift (hunk 1 above)
Comment (yusuftor) [reply Examples/SuperwallUIKitExample/SuperwallUIKitExample/Services/PaywallService.swift:26]: This is in the UIKit sample app, not the SDK haha

File: Sources/Paywall/Paywall/Paywall View Controller/SWWebView.swift [hunk 4]
```diff
@@ -61,3 +70,39 @@ final class SWWebView: WKWebView {
     fatalError("init(coder:) has not been implemented")
   }
 }
+
+// MARK: - WKNavigationDelegate
+extension SWWebView: WKNavigationDelegate {
+  func webView(
```
[the remark this answers is not in the document]
Comment (yusuftor) [reply Sources/Paywall/Paywall/Paywall View Controller/SWWebView.swift:76]: The default decision for a web view without implementing this method is actually `allow`. So people will have been able to visit links previously. I'll double check this now with a paywall link.

[the remark this answers is not in the document]
Comment (yusuftor) [reply Sources/Paywall/Paywall/Paywall View Controller/SWWebView.swift:76]: Check the latest commit, this fixes navigation

File: Sources/Paywall/Paywall/PaywallLogic.swift [hunk 5]
```diff
@@ -30,17 +30,28 @@ enum PaywallLogic {
...
 
+  static func trackAppInstall(
+    trackEvent: (Trackable) -> TrackingResult = Paywall.track
+  ) {
+    let appInstall = SuperwallEvent.AppInstall()
+    guard UserDefaults.standard.bool(forKey: appInstall.rawName) == false else {
+      return
+    }
+    _ = trackEvent(appInstall)
+    UserDefaults.standard.set(true, forKey: appInstall.rawName)
```
[the remark this answers is not in the document]
Comment (yusuftor) [reply Sources/Paywall/Paywall/PaywallLogic.swift]: Have updated with new app_install logic. Due to the queues in the cache, the tests mess up so I commented them out for now. I tested on device and confirmed the app_install works as expected. I wasn't sure what you meant by "change the timestamp of the event"... I've passed an "application_installed_at" parameter with the event though?

File: Sources/Paywall/Debug/SWDebugManager.swift (hunk 2 above)
Comment (yusuftor) [reply Sources/Paywall/Debug/SWDebugManager.swift]: Done

File: Sources/Paywall/Paywall/PaywallLogic.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (seth35us) [reply Sources/Paywall/Paywall/PaywallLogic.swift]: I noticed a memory leak on initPaywall() following the example in the [docs](https://docs.superwall.com/docs/configuring-the-sdk) for SwiftUi. Is that related to this: https://stackoverflow.com/questions/26383031/wkwebview-causes-my-view-controller-to-leak? Is there anything we can do to fix the leak?

[the remark this answers is not in the document]
Comment (yusuftor) [reply Sources/Paywall/Paywall/PaywallLogic.swift]: @seth35us I can look into this tomorrow but can you file an issue with more info about the leak you're experiencing?

**Pull request merged**